emcache/cluster.py

```diff
@@ -221,6 +254,80 @@ def node(self, memcached_host_address: MemcachedHostAddress) -> Node:
 
         raise ValueError(f"{memcached_host_address} can not be found")
 
+    async def _autodiscovery_monitor(self) -> None:
+        logger.debug("Autodiscovery task started")
+        while True:
+            if not self._autodiscovery:
+                break
+
+            try:
+                await asyncio.sleep(self._autodiscovery_poll_interval)
+                await self.autodiscover()
+            except asyncio.CancelledError:
+                # if a cancellation is received we stop monitoring for cluster changes
+                break
+
+            except Exception:
+                logger.exception("Autodiscover raised an exception, continuing processing")
+
+        logger.debug("Autodiscovery task stopped")
+
+    async def _get_autodiscovered_nodes(self) -> tuple[bool, int, list[tuple[str, str, int]]]:
+        if self._closed:
+            raise RuntimeError("Emcache client is closed")
+
+        node = random.choice(self._original_nodes)
+        async with OpTimeout(self._autodiscovery_timeout, self._loop):
+            async with node.connection() as connection:
+                return await connection.autodiscovery()
+
+    async def autodiscover(self) -> bool:
```

Comment:
Its been a while since I worked with Python and Asyncio, but could we have an instance variable for storing the result of the first success autodiscovery execution? With that we could avoid having to run multiple `autodiscovery` commands [here](https://github.com/emcache/emcache/pull/98/files#diff-d5918e537eadb7095bd9b366b7a19681b87d8d0f87683d5d4007a6fa4d5e861cR720) and [here](https://github.com/emcache/emcache/pull/98/files#diff-671bcd2cafefff8bcd836e45326e78a939d6c6276631c3976a8b5967fd27a54aR282)

IIRC we could do something like:

```python
self._first_autodiscovery_done = loop.create_future()
```

And then we could set the value when the first `autodiscovery` function would be successfully executed, something like:

```python
if not self._first_autodiscovery_done.done():
  self._first_autodiscovery_done.set_result(true)
```

And then later [here](https://github.com/emcache/emcache/pull/98/files#diff-d5918e537eadb7095bd9b366b7a19681b87d8d0f87683d5d4007a6fa4d5e861cR720) we could do something like:

```python
if autodiscovery:
  await asyncio.wait_for(client._cluster._first_autodiscovery_done, DEFAULT_STARTUP_WAIT_AUTODISCOVERY)
```

Thoughts?

Reply:
I updated the code to work that way. Yeah, it's better.